src/Controller/AuthenticationQrController.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types = 1);
+
+/**
+ * Copyright 2018 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\Tiqr\Controller;
+
+use InvalidArgumentException;
+use Psr\Log\LoggerInterface;
+use Surfnet\GsspBundle\Service\AuthenticationService;
+use Surfnet\GsspBundle\Service\StateHandlerInterface;
+use Surfnet\Tiqr\Tiqr\TiqrServiceInterface;
+use Surfnet\Tiqr\WithContextLogger;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Attribute\Route;
+
+class AuthenticationQrController
+{
+    public function __construct(
+        private readonly AuthenticationService $authenticationService,
+        private readonly StateHandlerInterface $stateHandler,
+        private readonly TiqrServiceInterface $tiqrService,
+        private readonly LoggerInterface $logger
+    ) {
+    }
+
+    /**
+     *
+     * @throws InvalidArgumentException
+     */
+    #[Route(path: '/authentication/qr', name: 'app_identity_authentication_qr', methods: ['GET'])]
+    public function __invoke(): Response
+    {
+        $nameId = $this->authenticationService->getNameId();
+        $sari = $this->stateHandler->getRequestId();
+        $logger = WithContextLogger::from($this->logger, ['nameId' => $nameId, 'sari' => $sari]);
+        $logger->info('Client request QR image');
+
+        // Do have a valid sample AuthnRequest?.
+        if (!$this->authenticationService->authenticationRequired()) {
+            $logger->error('There is no pending authentication request from SP');
+
+            return new Response('No authentication required', Response::HTTP_BAD_REQUEST);
```

Comment:
I've noticed that this code already existed before so it might be out of scope, but it seems to me that the status code of 400 is not correct here.

The request itself seems correct, however it can't be processed on the server. Not sure which status code is best here (because I lack the domain knowledge of when this happens in the grand picture), but it looks to me that the request itself seems valid.

Reply:
The request in this instance is not valid, the `authentication/qr` route is called without a pending authentication request present. This indicates a state issue, or someone is manually trying to open the route without having the correct preconditions (there must be a running authentication). So I think a 400 response is not incorrect. 

Forbidden (403) also seems fitting. I love these kind of discussions. This one I tend to leave as-is.

Reply:
If the reason is that the server is in an unexpected state, it sounds like [Precondition Failed](https://http.dev/412) (412) is a great fit.

400 should really be used if the request itself is incorrect (malformed, for example).

But I'm fine with leaving it as-is :).